java/gradle.java/src/org/netbeans/modules/gradle/java/api/GradleJavaSourceSet.java

```diff
@@ -162,7 +165,36 @@ public String getTargetCompatibility() {
      * @return the defined target compatibility
      */
     public String getTargetCompatibility(SourceType type) {
-        return targetCompatibility.getOrDefault(type, getSourcesCompatibility(type));
+        return fixCompatibility(type, targetCompatibility).orElse(getSourcesCompatibility(type));
+    }
+
+    /**
+     * Use compiler arguments to override source/target compatibility for JAVA.
+     * Look for something like "-flag" or "--flag" in args, the last occurence;
+     * return  it if found.
+     * <br>
+     * For example for source, flag is "-source", and args is
+     * "... --source 13 ..." then return "13". If not in args
+     * then don't change the compatibility, return the current value.
+     */
+    private Optional<String> fixCompatibility(SourceType sourceType, Map<SourceType, String> compatibilityMap) {
+        String compatibility = compatibilityMap.get(sourceType);
+        if(sourceType == SourceType.JAVA) { // only fixup for java
+            List<String> args = getCompilerArgs(sourceType);
+            if(args != null && !args.isEmpty()) {
+                String flag = compatibilityMap == sourcesCompatibility ? "-source" : "-target";
```

Comment:
Nitpick: maybe pass the flag source/target explicitly rather than comparing Map instances ?

Reply:
The first implementation did that (the one that was in the wrong place). I'll make that change after any other suggested changes have time to roll in. Thanks for the comment.

Reply:
Well, I'd reduce the number of ifs: getCompilerArgs() never returns with `null`, so line 184 could go. The double if in line 188 and 190 could be combined...

Reply:
I remember now why I compared to map instance; because -source/-target are/might-be specific to java. I'll change the method name to `fixJavaCompatibility`. I'd like to take out the souceType argument, but then the check for JAVA would, would have to be in the caller.

Reply:
K